### ISSUE TITLE
`require.main === module` is different after ncc (again)

### ISSUE THREAD

**Issue by KengoTODA** [vercel/ncc#748]
Thanks for sharing great tool to minify TS code, I'm using ncc so heavily for my GitHub Actions projects :)

Today I found an unexpected behavior same to the #224. I confirmed that v0.27.0 works fine but v0.28.0 doesn't, so there is a regression during [0.27.0...0.28.0](https://github.com/vercel/ncc/compare/0.27.0...0.28.0) I guess.

I've investigated about why this regression comes again even though #229 added a unit test?
It seems that [v0.14 has the test case](https://github.com/vercel/ncc/tree/0.14.0/test/unit) but [v0.15 does not have it](https://github.com/vercel/ncc/tree/0.15.0/test/unit), it seems that 8abb4ba071a483cdefc73d092c5151194e96dd57 (#264) removed the test case.

**Comment by KengoTODA** [comment]
Here is the [PR fixing problem in my repo](https://github.com/KengoTODA/readthedocs-action/pull/671/files#diff-3d2b59189eeedc2d428ddd632e97658fe310f587f7cb63b01f9b98ffc11c0197). It downgraded ncc from v0.29.0 to v0.27.0.

v0.29.0 uses `__nccwpck_require__.c[__nccwpck_require__.s] === module` which does not work, and
v0.27.0 uses `require.main === require.cache[eval('__filename')]` which works fine.

`__nccwpck_require__.c[__nccwpck_require__.s]` is an object like `{ id: 930, loaded: false, exports: Object [Module] { run: [Getter] } }`, and `module` is an object like `{ children: [], exports: [Setter] }`

Point is that, ncc can build the minimal case `console.log(require.main === module);` without problem. It generates `console.log(require.main === require.cache[eval('__filename')]);` just like v0.27.0. So maybe there is hidden conditions to use `__nccwpck_require__.c[__nccwpck_require__.s] === module` but I'm still not so sure.

**Comment by guybedford** [comment]
> Point is that, ncc v0.29.0 can build the minimal case console.log(require.main === module); without problem

I can verify that:

test.js
```js
console.log(require.main === module);
```

with `ncc run test.js` works just fine as you say.

So the question then is - under what conditions are you finding this fails?

Looking at your code you are running this check in an ES module file - that may actually be the issue here in that `require.main` and `module` aren't supposed to be defined in ES modules.

There is a way to do this same check in ES modules in Node.js although it is a little bit complex:

```js
import { pathToFileURL } from 'url';
import process from 'process';
if (import.meta.url === pathToFileURL(process.argv[1]).href)
  console.log('MAIN');
```

I just tested the above pattern in ncc, and unfortunately it appears that doesn't seem to be supported.

In that case I'd suggest we treat this issue as a feature request for the above pattern to be supported rather.

**Comment by guybedford** [comment]
(the test is still present and working just fine at https://github.com/vercel/webpack-asset-relocator-loader/tree/main/test/unit/require-main)

**Comment by guybedford** [comment]
The reason `import.meta` doesn't work here is that there doesn't seem to be a way to run a build that exposes `import.meta` to the output file - webpack always substitutes it even when doing a module build output.

@sokra are there any plans or ways to permit pass-through of `import.meta` for this kind of use case when outputting modules?

**Comment by KengoTODA** [comment]
Thank you! Your comment helped me a lot to clarify what is happening.
But I'm still not so clear about the problem, could you help me to understand:

How you found that the project is built as an ES module?
I think there is no related configuration in both [tsconfig.json](https://github.com/KengoTODA/readthedocs-action/blob/22524412ada4bc8e190839972e227deb772e48d4/tsconfig.json#L5) and [package.json](https://github.com/KengoTODA/readthedocs-action/blob/22524412ada4bc8e190839972e227deb772e48d4/package.json)... 🤔 

Thanks again for your support!

**Comment by KengoTODA** [comment]
According to the [`readme.md`](https://github.com/vercel/ncc/blob/37ccb67d6461ba8a850df3ca20a9ff0e0e760d1a/readme.md#usage-1):

> If building an .mjs or .js module inside a "type": "module" package boundary, an ES module output will be created automatically.

But in my project there is no `"type": "module"` in package.json.

**Comment by styfle** [comment]
@KengoTODA I think the TS bug is fixed in [0.31.0](https://github.com/vercel/ncc/releases/tag/0.31.0).

Please give it a try, thanks!

**Comment by KengoTODA** [comment]
Thanks @styfle, I've confirmed `@vercel/ncc` v0.31.0 generates `require.main === require.cache[eval('__filename')]` as expected!

I'll close this issue. @guybedford @styfle Thank you again :) 🙌
